src/java.base/unix/native/libjava/TimeZone_md.c

```diff
@@ -555,58 +555,20 @@ findJavaTZ_md(const char *java_home_dir)
 /**
  * Returns a GMT-offset-based zone ID. (e.g., "GMT-08:00")
  */
-
-#if defined(MACOSX)
-
 char *
 getGMTOffsetID()
 {
-    time_t offset;
-    char sign, buf[32];
-    struct tm local_tm;
-    time_t clock;
-
-    clock = time(NULL);
-    if (localtime_r(&clock, &local_tm) == NULL) {
-        return strdup("GMT");
-    }
-    offset = (time_t)local_tm.tm_gmtoff;
-    if (offset == 0) {
+    char buf[32];
+    char offset[6];
+    struct tm localtm;
+    time_t clock = time(NULL);
+    if (localtime_r(&clock, &localtm) == NULL) {
         return strdup("GMT");
     }
-    if (offset > 0) {
-        sign = '+';
-    } else {
-        offset = -offset;
-        sign = '-';
-    }
-    sprintf(buf, (const char *)"GMT%c%02d:%02d",
-            sign, (int)(offset/3600), (int)((offset%3600)/60));
-    return strdup(buf);
-}
 
-#else
+    strftime(offset, 6, "%z", &localtm);
+    char gmt_offset[] = {offset[0], offset[1], offset[2], ':', offset[3], offset[4], '\0'};
 
-char *
-getGMTOffsetID()
-{
-    time_t offset;
-    char sign, buf[32];
-    offset = timezone;
-
-    if (offset == 0) {
-        return strdup("GMT");
-    }
-
-    /* Note that the time offset direction is opposite. */
-    if (offset > 0) {
-        sign = '-';
-    } else {
-        offset = -offset;
-        sign = '+';
-    }
-    sprintf(buf, (const char *)"GMT%c%02d:%02d",
-            sign, (int)(offset/3600), (int)((offset%3600)/60));
+    sprintf(buf, (const char *)"GMT%s", gmt_offset);
```

Comment:
You could use the format string as "GMT%c%c%c:%c%c" so that the extra `gmt_offset` variable can be eliminated.

Reply:
Adjusted, forgot that the final string returned was able to be formatted itself as well.